### PR TITLE
Remove makefile

### PR DESCRIPTION
**Pull request by coldtobi**
This is how it looks like when removing the Makefile (made an extra branch so it is not entangled)

Missing is updating README.md

PS: I've added "DISTRIBUTION_VERSION" to the version so that distributions can ammend the version, which is maybe useful to determine where a bug report came from. Though there is no automatic detection, distribution needs to add that information in their build process)

### CODE REVIEW

File: man/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  )
+
+  add_custom_target(man ALL DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/timg.1)
+
+  install(FILES ${CMAKE_CURRENT_BINARY_DIR}/timg.1
+    DESTINATION ${CMAKE_INSTALL_FULL_MANDIR}/man1
+  )
+else()
+  message(WARNING "NOT re-generating the manpage. Possibly ronn is not installed.")
+  install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/timg.1.gz
```
Comment (hzeller) [inline man/CMakeLists.txt:18]: should we have the file distributed also be a non-gzipped form ? Or will this create confusion about using the generated target vs. version from source tree ?

**Comment by hzeller** [comment]
I think this looks great.
Yes, README update needed and should we now distribute the non-gzipped version of the manpage, to make things less confusing ?

**Comment by hzeller** [comment]
FYI, had to push some change in the Makefile; which doesn't matter as we remove it anyway in this pull request, but it will likely complain about a conflict.
New file is `renerer.o` and bumped C++ to 14, as it uses [`std::make_unique`](https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique) which probably can be described in the required features, though I haven't found [something that looks like that](https://cmake.org/cmake/help/latest/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html).

**Comment by hzeller** [comment]
Further changes since then: @speedy-beaver changed the manpage to use pandoc in #41

**Comment by coldtobi** [comment]
Great!
I've rebased this branch and will update with pandoc.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -1,13 +1,25 @@
...
+  thread-pool.h
+  video-display.h
+)
 
+target_link_libraries(timg PkgConfig::GRAPHICSMAGICKXX Threads::Threads)
 target_include_directories(timg PRIVATE ${CMAKE_BINARY_DIR}/src)
 
-# instead of just telling cmake "c++11, please", this a more modern approach:
-# https://cmake.org/cmake/help/latest/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html
-# https://crascit.com/2015/03/28/enabling-cxx11-in-cmake/ "Setting the C++ standard based on features"
```
Comment (coldtobi) [inline src/CMakeLists.txt]: The cmake page linked abe actually said that my knowledge of this cxx features is already outdated: 
`"These individual features are now less relevant and projects should generally prefer to use the high level meta features instead. Individual compile features are not provided for C++ 17 or later."
`
So I did just specify language level…

**Comment by coldtobi** [comment]
Update README.md
@speedy-beaver: Can you check if the MacOS instructions make sense? TIA!

File: README.md [hunk 3]
```diff
@@ -29,38 +29,49 @@ it).
...
+cd build
+LDFLAGS=-L${HOMEBREW_PREFIX}/lib cmake -DWITH_VIDEO_DECODING=On # or Off for no video
+make
 ```
 
 #### Install
 
-You can run timg directly in the source directory using `./timg`. To install
-the binary and [manpage](man/timg.1.md) on your system, type:
+You can run timg directly in the build directory using `./timg`. To install
```
Comment (hzeller) [inline README.md:73]: wouldn't that be `./src/timg` if you're in the build-directory ?

**Comment by coldtobi** [comment]
Am 9. Februar 2021 20:24:06 MEZ schrieb Henner Zeller <notifications@github.com>:
>@hzeller commented on this pull request.
>
>
>
>>  ```
> 
> #### Install
> 
>-You can run timg directly in the source directory using `./timg`. To
>install
>-the binary and [manpage](man/timg.1.md) on your system, type:
>+You can run timg directly in the build directory using `./timg`. To
>install
>
>wouldn't that be `./src/timg` if you're in the build-directory ?

your're right. i'll fix it later, (currently afk)

**Comment by speedy-beaver** [comment]
@coldtobi - reviewed the macOS instructions. With cmake the LDFLAGS are not needed, at least not when I run it on my machine. I suggest to change the text to this - also adding the `pandoc` dependency in the `brew install` command:

#### Compiling on macOS
```bash
# Homebrew needs to be available to install required dependencies
brew install GraphicsMagick webp cmake pandoc # required libs

# If you want to include video decoding, install these additional libraries
brew install ffmpeg

cmake .. -DWITH_VIDEO_DECODING=On # or Off for no video
make
```

If we want to deduplicate the Readme.md even more, then we can even have common text for `cmake.. ; make` steps.

File: README.md [hunk 4]
```diff
@@ -29,38 +29,49 @@ it).
...
-brew install GraphicsMagick webp  # required libs
+brew install GraphicsMagick webp cmake # required libs
 
 # If you want to include video decoding, install these additional libraries
 brew install ffmpeg
 
-LDFLAGS=-L${HOMEBREW_PREFIX}/lib make WITH_VIDEO_DECODING=1 # or 0 for no video
+mkdir build
+cd build
+LDFLAGS=-L${HOMEBREW_PREFIX}/lib cmake -DWITH_VIDEO_DECODING=On # or Off for no video
```
Comment (hzeller) [inline README.md:67]: this misses the `../` part in the call to cmake.

File: README.md [hunk 5]
```diff
@@ -29,38 +29,49 @@ it).
...
 
 # If you want to include video decoding, also install these additional libraries
-sudo apt-get install pkg-config libavcodec-dev libavformat-dev libswscale-dev
+sudo apt install libavcodec-dev libavformat-dev libswscale-dev
 
-make WITH_VIDEO_DECODING=1   # or 0, if you don't want video decoding
+# If you want to recreate the man page
+sudo apt install pandoc
+
+cmake ..
```
Comment (hzeller) [inline README.md:50]: Maybe we can make this similar in both install descriptions with one cmake line but that already contains the parameter (like below in the :apple:  installation)

```bash
cmake ../ -DWITH_VIDEO_DECODING=On  # of Off to for no video
```

Here we use `On` even though not needed, but we make it simple to discover that there is this setting. This way, we consume less attention span for someone not having to read the following two lines of comments :)

File: src/Makefile [hunk 6]
```diff
@@ -21,7 +21,11 @@ timg : $(OBJECTS)
 timg.o : timg-version.h
```
Comment (hzeller) [inline src/Makefile:21]: Wait, wouldn't this Makefile be removed ? Maybe this snug back in while rebasing ?
Also, the Makefile in `man/` can go as well now as we now generate it with cmake.

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  terminal-canvas.cc
+  timg.cc
+  display-options.h
+  image-source.h
+  terminal-canvas.h
+  timg-time.h
+  image-display.h
+  renderer.h
+  thread-pool.h
+  video-display.h
```
Comment (hzeller) [inline src/CMakeLists.txt:15]: These looks somewhat in random order, which makes it hard to add things later and discover duplicates.
I suggest to sort them (As exception maybe place `timg.cc` at the top still to indicate that this contains `main()`).

**Review by hzeller — CHANGES_REQUESTED** [review]
Mostly looks good, some comments below.

The details of how exactly to phrase things in the README, we can also leave up to some polishing later, so I guess for now it is only important that the LDFLAG part is not needed and that we need to call cmake with ../ in the macOS instructions.

File: src/Makefile (hunk 6 above)
Comment (coldtobi) [reply src/Makefile]: Yepp.. Likely. However, they are gone now.

File: README.md (hunk 4 above)
Comment (coldtobi) [reply README.md]: Fixed.
I've consolidated the build instructions, merging the identical bits for Debian and MacOS…

File: src/CMakeLists.txt (hunk 7 above)
Comment (coldtobi) [reply src/CMakeLists.txt]: yeah, I've just used ls' output… sorted now with timg.cc on top.
I've moved video-display.h as well, it should be in guarded by the if(WITH_VIDEO_DECODING)…

**Review by hzeller — APPROVED** [review]
Wonderful, thank you!

**Comment by hzeller** [comment]
Thanks, merged!

**Comment by hzeller** [comment]
I just pushed some minor change with a missing option in the pandoc generation, and re-added optimization and warning flags in the compilation.

One problem I have found: without git (e.g. if someone downloaded and compiled from https://github.com/hzeller/timg/archive/main.zip ), we don't get a fallback version string, but `timg --version` shows
```
timg HEAD-HASH-NOTFOUND HEAD-HASH-NOTFOUND <http://timg.sh/>
```
as version string.

Can you have a look how to make the fallback work @coldtobi ?

**Comment by coldtobi** [comment]
@hzeller Fixed in #42

**Pull request merged**